Jetpack blocks: isolate package

sirreal opened the pull request:
An intermediate step to isolate the package before moving it to Jetpack.
This will help to highlight dependencies.

This is not intended to be a long-lived, package. Indeed, at this point it cannot be completely isolated (it pulls in the Calypso root build). Soon, the blocks source will be moved out of Calypso and the package will not need to exist.

~Blocked: lib/format-currency handled here: #31229~ _Unblocked!_
Closes https://github.com/Automattic/wp-calypso/issues/31205
Closes https://github.com/Automattic/wp-calypso/issues/31209

## Testing
- npm run distclean
- npm ci
- Does Calypso build?
  Yes! Test the Gutenberg editor extensively!
  ```
  DISABLE_FEATURES=calypsoify/iframe npm start
  ```
- Do the block build?
  They should. Packages (including `jetpack-blocks`) will be built on install (`npm ci`).
  You can run the following to build just `jetpack-blocks`:
  ```sh
  SOURCEMAP='source-map' npx lerna run build --stream --scope '**/jetpack-blocks' -- -- --watch
  ````
  You'll need to sync `packages/jetpack-blocks/dist/` to `_inc/blocks/` for your Jetpack site. Be sure to remove stale files for testing. I'm using `rsync -a --delete packages/jetpack-blocks/dist/ ../jetpack/_inc/blocks/` for this.
- Do the blocks work well on the frontend of your Jetpack site? It's important to verify that the view scripts work well.

It's especially important to review things that depend on dynamic imports and the public path. They are:
- Simple Payments block in the editor (images in the preview view)
- Map block (editor/view)
- Slideshow block (editor/view)

sirreal committed:
Move package naively (broken)

sirreal committed:
Add eslint

sirreal committed:
Pull in override from root

sirreal committed:
Move preset into preset

sirreal committed:
Move blocks

sirreal committed:
Move utils and shared

sirreal committed:
Update build script

sirreal committed:
Move the stuff

sirreal committed:
Add dep to trigger errors

sirreal committed:
Add deps

sirreal committed:
Move jetpack-logo

sirreal committed:
DROPME: Don't resolve client imports

sirreal committed:
Drop calypso site fragment

sirreal committed:
DROPME: teach builder about package

sirreal committed:
Update src dir

sirreal committed:
Fix imports

sirreal committed:
Add deps

sirreal committed:
Add deps

sirreal committed:
Fix utils path

sirreal committed:
Update paths

sirreal committed:
Path fixing

sirreal committed:
Duplicate simple payments constants in block code

Temporary duplication of this code is acceptable to separate blocks from
Calypso.

sirreal committed:
DANGER! Import lib/format-currency outside package

This is not intended to be merged, but is a workaround until #31229
lands.

It should be replaced by '@automattic/format-currency' at that point.

sirreal commented on packages/jetpack-blocks/src/blocks/simple-payments/utils.js:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Internal dependencies
+ */
+import { getCurrencyDefaults } from '../../../../../client/lib/format-currency/currencies';
```
Not OK! Needs #31229 
```suggestion
import { getCurrencyDefaults } from '@automattic/format-currency/currencies';
```

ockham replied:
Fixed by the rebase.

ockham replied:
3d6cb23109

sirreal committed:
Pull in specific build

sirreal commented on bin/sdk/gutenberg.js:
```diff
@@ -6,12 +6,12 @@
 const fs = require( 'fs' );
```
Revert all this, pulled into specific block build

sirreal commented on packages/jetpack-blocks/bin/build.js:
```diff
@@ -0,0 +1,146 @@
+/**
+ * External dependencies
+ */
+const chalk = require( 'chalk' );
```
We can further simplify this file, no need to indirect through preset?

sirreal commented on packages/jetpack-blocks/bin/build.js:
```diff
@@ -0,0 +1,146 @@
+}
+
+const getBaseConfig = ( options = {} ) => {
+	const getConfig = require( path.join( '..', '..', '..', 'webpack.config.js' ) );
```
NOT OK! We need access to a base build package.

sirreal commented on packages/jetpack-blocks/src/editor-shared/get-site-fragment.js:
```diff
@@ -19,13 +9,6 @@ export default function getSiteFragment() {
-	if ( ! includes( [ 'post.php', 'post-new.php' ], siteFragment ) ) {
-		return siteFragment || null;
-	}
-
```
It's all Calypso-specific. We're moving out of —and no longer building specifically for— Calypso.

Makes https://github.com/Automattic/wp-calypso/issues/31209 irrelevant

sirreal replied:
If we restore this, we could land this while Gutenberg remains in Calypso by replacing the Calypso imports with `@automattic/jetpack-blocks/…` 🤔

sirreal commented on packages/jetpack-blocks/src/blocks/simple-payments/utils.js:
```diff
@@ -0,0 +1,29 @@
+/**
```
Duplicated here from lib.

Context: https://github.com/Automattic/wp-calypso/issues/31205

sirreal commented on packages/jetpack-blocks/src/blocks/simple-payments/constants.js:
```diff
@@ -0,0 +1,39 @@
+export const SIMPLE_PAYMENTS_PRODUCT_POST_TYPE = 'jp_pay_product';
```
Duplicated here from lib.

Context: https://github.com/Automattic/wp-calypso/issues/31205

ockham commented on packages/jetpack-blocks/package.json:
```diff
@@ -0,0 +1,63 @@
+    "photon": "^2.0.1",
+    "refx": "^3.1.1",
+    "resize-observer-polyfill": "^1.5.1"
+  },
```
I think there might be a few more dependencies that we could move over from the root `package.json` since they're only used by blocks, e.g. `mapbox-gl` (?)

sirreal replied:
There are! I just haven't gotten to them. The linter highlights them, mostly I lean on that to discover and pull them with a `^` from the root.

ockham commented:
Calypso currently doesn't start because

```
ERROR in ./client/gutenberg/editor/main.jsx
Module not found: Error: Can't resolve '../extensions/presets/jetpack/utils/refresh-registrations' in '/home/bernie/src/wp-calypso/client/gutenberg/editor'
```

and

```
ERROR in ./client/gutenberg/editor/controller.js
Module not found: Error: Can't resolve 'gutenberg/extensions/presets/jetpack/utils/get-jetpack-data' in '/home/bernie/src/wp-calypso/client/gutenberg/editor'
```

and

```
ERROR in ./client/blocks/inline-help/popover.jsx
Module not found: Error: Can't resolve 'gutenberg/extensions/presets/jetpack/utils/i18n' in '/home/bernie/src/wp-calypso/client/blocks/inline-help'
```

The last one is weird :slightly_frowning_face:

ockham commented:
Blocks won't currently build because

```
ERROR in Entry module not found: Error: Can't resolve '/home/bernie/src/wp-calypso/client/gutenberg/extensions/presets/jetpack/editor.js' in '/home/bernie/src/wp-calypso'
```

ockham commented:
Other than those errors, the code is looking pretty good :+1:

sirreal commented:
> Calypso currently doesn't start because

This is expected and preferred. With some changes dropping Gutenberg in Calypso support, we'll know when it's safe to proceed because the Calypso build errors will Disappear.

> ERROR in ./client/blocks/inline-help/popover.jsx
> Module not found: Error: Can't resolve 'gutenberg/extensions/presets/jetpack/utils/i18n' in '/home/bernie/src/wp-calypso/client/blocks/inline-help'
> ```
> The last one is weird 🙁

Yes, I was very curious where that came from - #29431 / @Copons 

https://github.com/Automattic/wp-calypso/blob/32d1c5797d58467c9cf879745a165009a0393751/client/blocks/inline-help/popover.jsx#L62

Should that be using `i18n-calypso` `translate` rather than `wp-i18n` `__`? It appears to be Calypso specific and likely outside the block editor.

---

> Blocks won't currently build because

😬 I'll update the blocks build instructions. It should be working, the package now has a `build` command that uses its specific webpack config. It will be run as part of `build-packages`.

sirreal committed:
Add deps

sirreal committed:
Drop format pragma

sirreal committed:
Use dist as build dir

sirreal committed:
Add @automattic/jetpack-blocks dep

sirreal committed:
Update jetpack-blocks paths in Calypso

sirreal committed:
Update-deps

sirreal committed:
Revert main gutenberg sdk changes

sirreal commented on babel.config.js:
```diff
@@ -56,7 +56,7 @@ const config = {
 	overrides: [
 		{
-			test: './client/gutenberg/extensions',
+			test: [ './client/gutenberg/extensions', './packages/jetpack-blocks' ],
```
I think this will work out because we're using Calypso's build right now (breaking out of the package boundary with `../../…`.

We'll need the build to get this right when it is published - #31281

sirreal committed:
Restore calypso-specific path handling

sirreal committed:
Add lib/format-currency FIXME

sirreal committed:
Fix snapshot tests

sirreal committed:
Already built

sirreal committed:
Fix ellipsis lint

sirreal committed:
Fix path

sirreal committed:
Fix more paths

sirreal committed:
No more gitignore

sirreal committed:
Clean clean script

ockham commented:
Imma rebase :+1:

ockham committed:
Add @automattic/format-currency dependency

ockham commented on packages/jetpack-blocks/src/blocks/simple-payments/edit.js:
```diff
@@ -23,13 +23,10 @@ import {
-} from 'lib/simple-payments/constants';
+import { __, _n } from '../../utils/i18n';
+import { decimalPlaces, formatPrice } from './utils';
+import { getCurrencyDefaults } from '../../../../../client/lib/format-currency/currencies';
```
Fixed by the rebase.

ockham committed:
Fix an import

ockham commented on packages/jetpack-blocks/src/editor-shared/get-site-fragment.js:
```diff
@@ -19,12 +20,14 @@ export default function getSiteFragment() {
 	if ( ! includes( [ 'post.php', 'post-new.php' ], siteFragment ) ) {
 		return siteFragment || null;
 	}
+	// FIXME: REMOVE TO HERE ^^^
```
Can we remove this stuff already now that Gutenlypso is out?

sirreal replied:
Gutelypso is still available, albeit behind a feature flag (disable `calypsoify/iframe`). Until it's really removed and inaccessible, I think we need to continue full support.

ockham commented:
Testing blocks in Jetpack, I'm getting `The Swiper library could not be loaded.` from the Slideshow block upon trying to add images.

sirreal committed:
Move swiper dep

sirreal committed:
Update deps

sirreal commented:
> Testing blocks in Jetpack, I'm getting The Swiper library could not be loaded. from the Slideshow block upon trying to add images.

I've seen this randomly and don't believe it's from any changes in this branch. It would be nice to understand why that happens, perhaps just from fragile network with ngrok and local Jetpack docker.

I've just tested and did not get this error. Do you get this error consistently?

ockham commented:
> > Testing blocks in Jetpack, I'm getting The Swiper library could not be loaded. from the Slideshow block upon trying to add images.
> 
> I've seen this randomly and don't believe it's from any changes in this branch. It would be nice to understand why that happens, perhaps just from fragile network with ngrok and local Jetpack docker.
> 
> I've just tested and did not get this error. Do you get this error consistently?

Yeah, seems consistent, and never getting it on `master`...

sirreal committed:
Import public-path not frontend management

With path moving, shared import was screwed up.

Fix it.

ockham commented:
Slideshow styling is still broken on the frontend:

![image](https://user-images.githubusercontent.com/96308/54121493-db33c980-43fa-11e9-8d93-e9c948fbb42d.png)

ockham commented:
Map block assets also aren't showing on the frontend

![image](https://user-images.githubusercontent.com/96308/54122912-a9bcfd00-43fe-11e9-9b93-2db9bb45595a.png)

sirreal commented:
Thanks @ockham 

That's the same issue as the editor problem fix in b4dc2b1 but for the view scripts 😞 

I'll take a look.

sirreal committed:
Fix editor/view setup

sirreal committed:
Update outdated webpackInclude

sirreal committed:
Fix attribute placeHolder -> placeholder

sirreal committed:
Fix view script paths

sirreal commented:
I've rebased fixed a bunch of issues:

- We were prepending the JS from some directories to webpack entries. I've removed that in favor of a few specific files: `preset/setup/editor.js` and `preset/setup/view.js`. These include the necessary setup for editor and view code. They only include the public path and the Jetpack category, a lot of extraneous code was being added to the entry.
- The view paths were wrong so no view scripts were being created. This is fixed.
- The dynamic imports in the Calypso editor had a stale `webpackInclude` comment which resulted in no blocks being added. This has been fixed.
- More things…

In my testing this is working well, I'd appreciate thorough testing and review 🙇

sirreal committed:
Production block build in CI

sirreal committed:
Build prod blocks on CI

sirreal committed:
Include dist

sirreal commented:
Some build stats for comparison. Run `npm ci`, cd to `packge.json` directory, `npm publish --dry-run`:

<table>
<thead><tr>
<th><code>master</code>
</th><th><code>branch</code></th></tr></thead>
<tbody>
<tr>
<td>

```
npm notice
npm notice 📦  @automattic/jetpack-blocks@13.1.1
npm notice === Tarball Contents ===
npm notice 906B    package.json
npm notice 508B    README.md
npm notice 51B     build/contact-info/view.css
npm notice 1.2kB   build/contact-info/view.js
npm notice 51B     build/contact-info/view.rtl.css
npm notice 66.8kB  build/editor-beta.css
npm notice 1.2MB   build/editor-beta.js
npm notice 66.8kB  build/editor-beta.rtl.css
npm notice 65.3kB  build/editor.css
npm notice 1.2MB   build/editor.js
npm notice 65.3kB  build/editor.rtl.css
npm notice 589B    build/gif/view.css
npm notice 1.2kB   build/gif/view.js
npm notice 590B    build/gif/view.rtl.css
npm notice 85.3kB  build/images/map-theme_black_and_white-1ead5946ca104d83676d6e3410e1d733.jpg
npm notice 113.5kB build/images/map-theme_default-2ceb449b599dbcbe2a90fead5a5f3824.jpg
npm notice 160.3kB build/images/map-theme_satellite-c74dc129bda9502fb0fb362bb627577e.jpg
npm notice 108.7kB build/images/map-theme_terrain-2b6e6c1c8d09cbdc58a4c0653be1a6e3.jpg
npm notice 1.2kB   build/images/oval-3cc7669d571aef4e12f34b349e42d390.svg
npm notice 7.5kB   build/images/paypal-button-1e53882e702881f8dfd958c141e65383.png
npm notice 8.2kB   build/images/paypal-button-2x-fe4d34770a47484f401cecbb892f8456.png
npm notice 361B    build/index.json
npm notice 689B    build/mailchimp/view.css
npm notice 2.9kB   build/mailchimp/view.js
npm notice 689B    build/mailchimp/view.rtl.css
npm notice 23.8kB  build/map/mapbox-gl.css
npm notice 655.6kB build/map/mapbox-gl.js
npm notice 23.8kB  build/map/mapbox-gl.rtl.css
npm notice 406B    build/map/view.css
npm notice 20.0kB  build/map/view.js
npm notice 407B    build/map/view.rtl.css
npm notice 3.5kB   build/repeat-visitor/view.js
npm notice 6.1kB   build/slideshow/view.css
npm notice 92.9kB  build/slideshow/view.js
npm notice 6.1kB   build/slideshow/view.rtl.css
npm notice 18.2kB  build/swiper.css
npm notice 127.7kB build/swiper.js
npm notice 18.3kB  build/swiper.rtl.css
npm notice 8.1kB   build/tiled-gallery/view.css
npm notice 12.3kB  build/tiled-gallery/view.js
npm notice 8.2kB   build/tiled-gallery/view.rtl.css
npm notice === Tarball Details ===
npm notice name:          @automattic/jetpack-blocks
npm notice version:       13.1.1
npm notice package size:  1.2 MB
npm notice unpacked size: 4.1 MB
npm notice shasum:        a4a0bb3fa4c108c1bb26baccde495c38310c141d
npm notice integrity:     sha512-psDpPmM6plC5j[...]rh2d609a8dPzw==
npm notice total files:   41
npm notice
```

</td>
<td>

```
npm notice
npm notice 📦  @automattic/jetpack-blocks@13.1.1
npm notice === Tarball Contents ===
npm notice 2.1kB   package.json
npm notice 508B    README.md
npm notice 51B     dist/contact-info/view.css
npm notice 1.3kB   dist/contact-info/view.js
npm notice 51B     dist/contact-info/view.rtl.css
npm notice 66.8kB  dist/editor-beta.css
npm notice 1.1MB   dist/editor-beta.js
npm notice 66.8kB  dist/editor-beta.rtl.css
npm notice 65.3kB  dist/editor.css
npm notice 1.1MB   dist/editor.js
npm notice 65.3kB  dist/editor.rtl.css
npm notice 589B    dist/gif/view.css
npm notice 1.3kB   dist/gif/view.js
npm notice 590B    dist/gif/view.rtl.css
npm notice 85.3kB  dist/images/map-theme_black_and_white-1ead5946ca104d83676d6e3410e1d733.jpg
npm notice 113.5kB dist/images/map-theme_default-2ceb449b599dbcbe2a90fead5a5f3824.jpg
npm notice 160.3kB dist/images/map-theme_satellite-c74dc129bda9502fb0fb362bb627577e.jpg
npm notice 108.7kB dist/images/map-theme_terrain-2b6e6c1c8d09cbdc58a4c0653be1a6e3.jpg
npm notice 1.2kB   dist/images/oval-3cc7669d571aef4e12f34b349e42d390.svg
npm notice 7.5kB   dist/images/paypal-button-1e53882e702881f8dfd958c141e65383.png
npm notice 8.2kB   dist/images/paypal-button-2x-fe4d34770a47484f401cecbb892f8456.png
npm notice 361B    dist/index.json
npm notice 689B    dist/mailchimp/view.css
npm notice 3.0kB   dist/mailchimp/view.js
npm notice 689B    dist/mailchimp/view.rtl.css
npm notice 23.8kB  dist/map/mapbox-gl.css
npm notice 655.6kB dist/map/mapbox-gl.js
npm notice 23.8kB  dist/map/mapbox-gl.rtl.css
npm notice 406B    dist/map/view.css
npm notice 16.2kB  dist/map/view.js
npm notice 407B    dist/map/view.rtl.css
npm notice 3.7kB   dist/repeat-visitor/view.js
npm notice 6.1kB   dist/slideshow/view.css
npm notice 19.7kB  dist/slideshow/view.js
npm notice 6.1kB   dist/slideshow/view.rtl.css
npm notice 18.2kB  dist/swiper.css
npm notice 127.7kB dist/swiper.js
npm notice 18.3kB  dist/swiper.rtl.css
npm notice 8.1kB   dist/tiled-gallery/view.css
npm notice 11.7kB  dist/tiled-gallery/view.js
npm notice 8.2kB   dist/tiled-gallery/view.rtl.css
npm notice === Tarball Details ===
npm notice name:          @automattic/jetpack-blocks
npm notice version:       13.1.1
npm notice package size:  1.2 MB
npm notice unpacked size: 3.9 MB
npm notice shasum:        f2cb54c74c9732373ba87e8ea6a4ceeb8cab02c1
npm notice integrity:     sha512-47gmy4VLdVONH[...]IaSbB7YGzbuPg==
npm notice total files:   41
npm notice
```

</td></tr>
</tbody>
</table>

sirreal committed:
Fix scope in CI build

gititon commented:
Thanks for the very detailed testing instructions, @sirreal. I checked edit and view of nearly all of the Jetpack blocks, paying special attention to Maps, Simple Payments, and Slideshow, and they all tested perfectly for me!

ockham reviewed (approved):
Code looks good, too :tada: :ship:

sirreal commented:
Thanks for the testing, folks!

I'll plan to land this tomorrow if no objections are raised.
